Mathlib/CategoryTheory/FiberedCategory/Fiber.lean

```diff
@@ -0,0 +1,110 @@
+/-
+Copyright (c) 2024 Calle Sönne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Calle Sönne, Paul Lezeau
+-/
+
+import Mathlib.CategoryTheory.FiberedCategory.HomLift
+import Mathlib.CategoryTheory.Functor.Const
+
+/-!
+
+# Fibers of a functors
+
+In this file we define, for a functor `p : 𝒳 ⥤ 𝒴`, the fiber categories `Fiber p S` for every
+`S : 𝒮` as follows
+- An object in `Fiber p S` is a pair `(a, ha)` where `a : 𝒳` and `ha : p.obj a = S`.
+- A morphism in `Fiber p S` is a morphism `φ : a ⟶ b` in 𝒳 such that `p.map φ = 𝟙 S`.
+
+For any category `C` equipped with a functor `F : C ⥤ 𝒳` such that `F ⋙ p` is constant at `S`,
+we define a functor `InducedFunctor : C ⥤ Fiber p S` that `F` factors through.
+-/
+
+universe v₁ u₁ v₂ u₂ v₃ u₃
+
+open CategoryTheory Functor Category IsHomLift
+
+variable {𝒮 : Type u₁} {𝒳 : Type u₂} [Category.{v₁} 𝒮] [Category.{v₂} 𝒳]
+
+/-- `Fiber p S` is the type of elements of `𝒳` mapping to `S` via `p`.  -/
+def Fiber (p : 𝒳 ⥤ 𝒮) (S : 𝒮) := {a : 𝒳 // p.obj a = S}
+
+namespace Fiber
+
+/-- `Hom a b` are the morphisms of `Fiber p S`, defined as those lying over `𝟙 S` in the base. -/
+def Hom {p : 𝒳 ⥤ 𝒮} {S : 𝒮} (a b : Fiber p S) := {φ : a.1 ⟶ b.1 // IsHomLift p (𝟙 S) φ}
+
+instance {p : 𝒳 ⥤ 𝒮} {S : 𝒮} (a b : Fiber p S) (φ : Hom a b) : IsHomLift p (𝟙 S) φ.1 := φ.2
+
+/-- `Fiber p S` has the structure of a category with morphisms being those lying over `𝟙 S`. -/
+@[simps]
+instance FiberCategory {p : 𝒳 ⥤ 𝒮} {S : 𝒮} : Category (Fiber p S) where
+  Hom a b := {φ : a.1 ⟶ b.1 // IsHomLift p (𝟙 S) φ}
+  id a := ⟨𝟙 a.1, IsHomLift.id a.2⟩
+  comp φ ψ := ⟨φ.val ≫ ψ.val, inferInstance⟩
```

Comment:
I would suggest the following design:
```suggestion
variable {p : 𝒳 ⥤ 𝒮} {S : 𝒮}

/-- `Fiber p S` has the structure of a category with morphisms being those lying over `𝟙 S`. -/
instance fiberCategory : Category (Fiber p S) where
  Hom a b := {φ : a.1 ⟶ b.1 // IsHomLift p (𝟙 S) φ}
  id a := ⟨𝟙 a.1, IsHomLift.id a.2⟩
  comp φ ψ := ⟨φ.val ≫ ψ.val, by have := φ.2; have := ψ.2; infer_instance⟩

@[simps obj]
def fiberInclusion : Fiber p S ⥤ 𝒳 where
  obj a := a.1
  map φ := φ.1

@[simp]
lemma fiberInclusion_map_id (a : Fiber p S) : fiberInclusion.map (𝟙 a) = 𝟙 _ := rfl

instance {a b : Fiber p S} (φ : a ⟶ b) :
    IsHomLift p (𝟙 S) (fiberInclusion.map φ) := φ.2

@[ext]
lemma hom_ext {a b : Fiber p S} {φ ψ : a ⟶ b}
    (h : fiberInclusion.map φ = fiberInclusion.map ψ) : φ = ψ :=
  Subtype.ext h

instance : (fiberInclusion : Fiber p S ⥤ _).Faithful where

lemma fiberInclusion_comp_const :
    fiberInclusion ⋙ p = (const (Fiber p S)).obj S :=
  Functor.ext (fun x ↦ x.2) (fun _ _ φ ↦ IsHomLift.fac' p (𝟙 S) (fiberInclusion.map φ))
```

Reply:
Then, `fiberInclusion.map φ` would become the canonical form, instead of `φ.2`.

Reply:
I have gone with this for now, but I am worried that making the parameters `p` and `S` implicit to `fiberInclusion` could mess with things later on (I seem to remember this happening when working on Stacks).